### ISSUE TITLE
Refactor of the loss into a model class

### ISSUE THREAD

**Issue by josephdviviano** [saleml/torchgfn#48]
Basically, the current way of interacting with the parameterization is a bit confusing to explain to others. We have parameters, which produce predictions, in conjunction with trajectories which are sampled, and this produces a loss. But when we call the loss function, we never see the parameters. We propose here to define a `model` with parameterizations, which has a `model.loss()` method, which accepts trajectories, to produce a loss.

Simmply:
```
# Define a model and a sampler.
tb_model = TBParametrization(logit_PF, logit_PB, logZ)
trajectories_sampler = TrajectoriesSampler(actions_sampler=actions_sampler)

# Get some trajectories and produce a loss.
trajectories = trajectories_sampler.sample(n_trajectories=16)
loss = tb_model.loss(trajectories)
```

More concretely,

```
    env = HyperGrid(ndim=4, height=8, R0=0.01)  # Grid of size 8x8x8x8

    module_PF = NeuralNet(input_dim=env.preprocessor.output_shape[0], output_dim=env.n_actions)
    module_PB = NeuralNet(input_dim=env.preprocessor.output_shape[0], output_dim=env.n_actions - 1, torso=module_PF.torso)

    logit_PF = DiscretePFEstimator(env=env, module=module_PF)
    logit_PB = DiscretePBEstimator(env=env, module=module_PB)
    logZ = LogZEstimator(torch.tensor(0.0))

    # In another PR let's consider merging these (`trajectory_sampler` becomes a 
    # method of `actions_sampler`).
    actions_sampler = ActionsSampler(estimator=logit_PF)
    trajectories_sampler = TrajectoriesSampler(actions_sampler=actions_sampler)

    # To remove.
    #parametrization = TBParametrization(logit_PF, logit_PB, logZ)
    #loss_fn = TrajectoryBalance(parametrization=parametrization)
    
    # Replaced with.
    tb_model = TBParametrization(logit_PF, logit_PB, logZ)

    params = [
        {
            "params": [
                val for key, val in parametrization.parameters.items() if "logZ" not in key
            ],
            "lr": 0.001,
        },
        {"params": [val for key, val in parametrization.parameters.items() if "logZ" in key], "lr": 0.1},
    ]
    optimizer = torch.optim.Adam(params)

    for i in (pbar := tqdm(range(1000))):
        trajectories = trajectories_sampler.sample(n_trajectories=16)
        optimizer.zero_grad()

        # loss = loss_fn(trajectories)  # replaced with
        loss = tb_model.loss(trajectories)

        loss.backward()
        optimizer.step()
        if i % 25 == 0:
            pbar.set_postfix({"loss": loss.item()})

```

This is largely a renaming of the currently available abstraction made to make the code easier to grok to newcomers.

**Comment by josephdviviano** [comment]
In a follow-up, we can explore whether making a trajectories sampler a method of the actions sampler, so we just have to define a single sampler.

Finally, we could explore a GFN class which accepts both a sampler and a model. But this would be purely for convenience.

**Comment by saleml** [comment]
This is handled in https://github.com/saleml/torchgfn/pull/89. Good job !
